src/nvim/channel.c

```diff
@@ -0,0 +1,740 @@
+    case kChannelStreamProc:
+      if (chan->stream.proc.type == kProcessTypePty) {
+          stream_desc = "pty job";
+      } else {
+          stream_desc = "job";
+      }
+      break;
+
+    case kChannelStreamStdio:
+       stream_desc = "stdio";
+       break;
+
+    case kChannelStreamSocket:
+      stream_desc = "socket";
+      break;
+
+    case kChannelStreamInternal:
+      stream_desc = "socket (internal)";
+      break;
+
+    default:
+      stream_desc = "?";
+  }
+
+  if (chan->is_rpc) {
+    mode_desc = ", rpc";
+  } else if (chan->term) {
+    mode_desc = ", terminal";
+  } else {
+    mode_desc = "";
+  }
+
+  if (ext_source) {
+    // TODO(bfredl): in a future improved traceback solution,
+    // external events should be included.
+    source = ext_source;
+  } else {
+    eval_format_source_name_line((char *)IObuff, sizeof(IObuff));
+    source = (char *)IObuff;
+  }
+
+  ILOG("new channel %" PRIu64 " (%s%s): %s", chan->id, stream_desc,
+       mode_desc, source);
+#else
+  (void)chan;
+  (void)ext_source;
+#endif
+}
+
+void channel_incref(Channel *channel)
+{
+  channel->refcount++;
+}
+
+void channel_decref(Channel *channel)
+{
+  if (!(--channel->refcount)) {
+    multiqueue_put(main_loop.fast_events, free_channel_event, 1, channel);
+  }
+}
+
+void callback_reader_free(CallbackReader *reader)
+{
+  callback_free(&reader->cb);
+  if (reader->buffered) {
+    ga_clear(&reader->buffer);
+  }
+}
+
+void callback_reader_start(CallbackReader *reader)
+{
+  if (reader->buffered) {
+    ga_init(&reader->buffer, sizeof(char *), 1);
+  }
+}
+
+static void free_channel_event(void **argv)
+{
+  Channel *channel = argv[0];
+  if (channel->is_rpc) {
+    rpc_free(channel);
+  }
+
+  callback_reader_free(&channel->on_stdout);
+  callback_reader_free(&channel->on_stderr);
+  callback_free(&channel->on_exit);
+
+  pmap_del(uint64_t)(channels, channel->id);
+  multiqueue_free(channel->events);
+  xfree(channel);
+}
+
+static void channel_destroy_early(Channel *chan)
+{
+  if ((chan->id != --next_chan_id)) {
+    abort();
+  }
+
+  if ((--chan->refcount != 0)) {
+    abort();
+  }
+
+  free_channel_event((void **)&chan);
+}
+
+
+static void close_cb(Stream *stream, void *data)
+{
+  channel_decref(data);
+}
+
+Channel *channel_job_start(char **argv, CallbackReader on_stdout,
+                           CallbackReader on_stderr, Callback on_exit,
+                           bool pty, bool rpc, bool detach, const char *cwd,
+                           uint16_t pty_width, uint16_t pty_height,
+                           char *pty_name, varnumber_T *status_out)
+{
+  Channel *chan = channel_alloc(kChannelStreamProc);
+  chan->on_stdout = on_stdout;
+  chan->on_stderr = on_stderr;
+  chan->on_exit = on_exit;
+  chan->is_rpc = rpc;
+
+  if (pty) {
+    if (detach) {
+      EMSG2(_(e_invarg2), "terminal/pty job cannot be detached");
+      shell_free_argv(argv);
+      channel_destroy_early(chan);
+      *status_out = 0;
+      return NULL;
```

Comment:
`pty_name` is not freed. Triggered by
`:echo jobstart('ls',{'pty':1,'detach':1,'TERM':"lsdjfslkdfjslkdfjsldkfklsdfsd"})`